aspnet/security/authorization/secure-data.rst

```diff
@@ -0,0 +1,369 @@
+
+Currently the UI shows edit and delete links for data the user cannot modify. We'll fix that by applying the authorization filter to the views.
+
+Inject the authorization service in the *Views/_ViewImports.cshtml* file so it will be available to all views:
+
+.. literalinclude:: secure-data/samples/final/Views/_ViewImports.cshtml
+  :language: none
+  :emphasize-lines: 7,8
+
+Update the *Views/Contacts/Index.cshtml* Razor view to show only display the edit and delete links for the users data:
+
+.. literalinclude:: secure-data/samples/final/Views/Contacts/Index.cshtml
+  :language: none
+  :lines: 2
+
+.. literalinclude:: secure-data/samples/final/Views/Contacts/Index.cshtml
+  :language: none
+  :emphasize-lines: 6-10, 12-17
+  :start-after: <snippet_1>
+  :end-before: </snippet_1>
+  :dedent: 16
+
+Test the app with two different browsers and users to verify they cannot edit or delete contacts they didn't create.
+
+.. warning:: Hiding links from users that do not have permission to edit or delete data does not secure that app, it makes the app more user friendly by displaying only valid links. Users can hack the generated URLs to invoke edit and delete operations on data they don't own, it's up to the controller to secure the data.
+
+Adding an administrative role
+-------------------------------
+
+Applications frequently support an administrator account with permissions to modify user data.  In this sample we'll create the "canDelete" role, and users in this role will be able to delete any contacts. A best practice is to name roles by the actions they can perform, so "canDelete" is preferred over a role called "admin". When your application evolves, you can add new roles such as "canDeleteMembers" rather than the less descriptive "superAdmin".
+
+The "canDelete" requirement will have two handlers, one for the contact owner and one for administrative users in the "canDelete" role. Using multiple handlers for a requirement significantly simplifies your code, as the controller and UI code don't change, you simply add the administrative "canDelete" filter. See :ref:`security-authorization-policies-based-authorization-handler` for more information.
+
+Log out of the "test@example.com" browser session or close the browser to clear the identity cookie.
+
+Update ``SeedData`` to create and call the "canDelete" role:
+
+.. literalinclude::  secure-data/samples/final/Data/SeedData.cs
+  :language: c#
+  :start-after:  #region snippet_CreateCanDeleteRole
+  :end-before: #endregion
+  :dedent: 8
+  
+.. literalinclude::  secure-data/samples/final/Data/SeedData.cs
+  :language: c#
+  :start-after:  #region snippet_Initialize
+  :end-before: #endregion
+  :emphasize-lines: 7
+  :dedent: 8
+  
+Create a new class called ``ContactRoleAuthorizationHandler`` for the role filter to validate the user is in the specified role:
+
+.. literalinclude::  secure-data/samples/final/Authorization/ContactRoleAuthorizationHandler.cs
```

Comment:
Then, if we have the admin role, you'd have two handlers for the requirement/operation, one which is "Admin" which ignores the operation entirely.